lumicks/pylake/tests/test_correlated_stack.py

```diff
@@ -119,3 +186,239 @@ def test_correlation():
+
+
+def make_alignment_image_data(spots, Tx_red, Ty_red, theta_red, Tx_blue, Ty_blue, theta_blue, bit_depth,
+                              offsets=None):
+    def make_transform_matrix(Tx, Ty, theta):
+        M = np.eye(3)
+        M[0, -1] = Tx
+        M[1, -1] = Ty
+        theta = np.radians(theta)
+        M[0, 0] = np.cos(theta)
+        M[0, 1] = -np.sin(theta)
+        M[1, 0] = np.sin(theta)
+        M[1, 1] = np.cos(theta)
+        return M
+
+    def transform_spots(M, spots):
+        # apply channel offsets
+        # reshape spots into coordinate matrix; [x,y,z] as columns
+        N = spots.shape[1]
+        spots = np.vstack((spots, np.ones(N)))
+        return np.dot(M, spots)[:2]
+
+    def make_image(spots_red, spots_green, spots_blue, bit_depth):
+        # RGB image, 2D (normalized) gaussians at spot locations
+        sigma = np.eye(2)*5
+        X, Y = np.meshgrid(np.arange(0, 200), np.arange(0, 100))
+        img = np.zeros((*X.shape, 3))
+        for j, pts in enumerate((spots_red.T, spots_green.T, spots_blue.T)):
+            for x, y, in pts:
+                mu = np.array([x,y])[:,np.newaxis]
+                XX = np.vstack((X.ravel(), Y.ravel())) - mu
+                quad_form = np.sum(np.dot(XX.T, np.linalg.inv(sigma)) * XX.T, axis=1)
+                Z = np.exp(-0.5 * quad_form) 
+                img[:, :, j] += Z.reshape(X.shape)
+            img[:, :, j] = img[:, :, j] / img[:, :, j].max()
+        return (img * (2**bit_depth - 1)).astype(f"uint{bit_depth}")
+
+    def make_description(m_red, m_blue, offsets):
+        if offsets is None:
+            offsets = [0, 0]
+        # WARP_INVERSE_MAP flag requires original transformation that resulted in un-aligned image
+        return {"Alignment red channel": m_red[:2].ravel().tolist(),
+                "Alignment green channel": np.eye(3)[:2].ravel().tolist(),
+                "Alignment blue channel": m_blue[:2].ravel().tolist(),
+                "Alignment region of interest (x, y, width, height)": [offsets[0], offsets[1], 200, 100],
+                "Region of interest (x, y, width, height)": [0, 0, 200, 100]}
+
+    spots = np.array(spots).T # [2 x N]
+    img0 = make_image(spots, spots, spots, bit_depth)
+
+    if offsets is not None: # translate origin by offsets
+        spots[0] -= offsets[0]
+        spots[1] -= offsets[1]
+    m_red = make_transform_matrix(Tx_red, Ty_red, theta_red)
+    m_blue = make_transform_matrix(Tx_blue, Ty_blue, theta_blue)
+    red_spots = transform_spots(m_red, spots)
+    blue_spots = transform_spots(m_blue, spots)
+    if offsets is not None: # back-translate origin
+        for tmp_spots in (red_spots, spots, blue_spots):
+            tmp_spots[0] += offsets[0]
+            tmp_spots[1] += offsets[1]
+    
+    img = make_image(red_spots, spots, blue_spots, bit_depth)
+    description = make_description(m_red, m_blue, offsets)
+
+    return img0, img, description            
+
+
+def test_image_reconstruction():
```

Comment:
This test feels like a bunch of small tests in a single big test.   This is not ideal since one test can observe side effects from another.

Maybe we could split them?
The comments that already indicate different things being tested are a great reference point.
Apparently, the suggested thing to group them is to put them in a class [docs](https://pytestguide.readthedocs.io/en/latest/pytestGuide/#test-code).

Same thing might apply to `test_rgb_correlated_stack`.

Reply:
Agreed. I left `test_rgb_correlated_stack` as it because it's the same format as `test_correlated_stack`. Now that the testing is growing, it might be worthwhile to reorganize the whole file in another ticket